x-pack/filebeat/input/awss3/s3_objects.go

```diff
@@ -233,6 +255,37 @@ func (p *s3ObjectProcessor) readJSON(r io.Reader) error {
 	return nil
 }
 
+// readJSONSlice uses a slice of json.RawMessage to process JSON slice data
+// as individual JSON objects.
+func (p *s3ObjectProcessor) readJSONSlice(r io.Reader) error {
+	dec := json.NewDecoder(r)
+	dec.UseNumber()
+
+	for dec.More() && p.ctx.Err() == nil {
+		offset := dec.InputOffset()
+
+		var items []json.RawMessage
+		if err := dec.Decode(&items); err != nil {
+			return fmt.Errorf("failed to decode json: %w", err)
+		}
+
+		for _, item := range items {
+			if p.readerConfig.ExpandEventListFromField != "" {
+				if err := p.splitEventList(p.readerConfig.ExpandEventListFromField, item, offset, p.s3ObjHash); err != nil {
+					return err
+				}
+				continue
+			}
+
+			data, _ := item.MarshalJSON()
+			evt := p.createEvent(string(data), offset)
```

Comment:
If this JSON array contains more than one event then they all will have the same offset. The `offset` is meant to hold the byte index of the first character of the JSON object.

If two events have the same offset then they have the same `_id` which means one of them will be discarded as a duplicate.

Reply:
@andrewkroh agree with this and have updated the logic to track offsets more effectively. I have concerns that this might not solve the issue in some cases. Since the parquet reader does not return a traditional offset and the io.Reader is created with every fresh byte slice returned, every slice batch that it returns will ideally start with the same offset when passed to the JSON decoder. i.e If there are 3000 records with each row and JSON object having the same size where for example batch size is 1000, then the 3 batches would have the same offset for all the objects even though each record in a 1000 batch have different offsets to each other. That would mean it would publish the first 1000 records and discard the rest 2000 ?

Reply:
Are events always returned in the same order from the parquet reader? If so, then perhaps instead of treating the offset as a byte index we use it as an event index.

Reply:
Yes the parquet reader always maintains the order of events .

Reply:
Treating the offset as an event index should work. Would anything need to change here or does it look good for merging for now ?

Reply:
added event offset tracking